### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] asserting empty object type to a type alias of `Record` false positive

### ISSUE THREAD

**Issue by hesprs** [typescript-eslint/typescript-eslint#12856]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=MYewdgzgLgBAhjAvDA3gXwNwCgsLhGAJQFNQAnAEwB5oyBLMAcwBoZaHGA%2BDGAel5gB1EGQDWBAGYNiOKAE8ADsRgAROsFjIS5auyas9XbHgJqNPfjABicADYRlCkBDpQ6ANxlA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWwMOIkrIBDaPmoFi1AcmSJoucmwbooM6G2iQAviA1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkAhqqrAO4BSZlemArugDTgakEl2TQ8AOW6MAwgAt00ANYcwXXoIC%2BITUA&tokens=false

### Repro Code

```TypeScript
const a = {};
a as Record<string, string>; // Works fine
type Dict = Record<string, string>;
a as Dict; // False positive
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

The type alias of a `Record` should not trigger `no-unnecessary-type-assertion`; align behavior with direct non-alias assertion.

### Actual Result

TsEslint flagged the fourth line `no-unnecessary-type-assertion`.

### Additional Info

Seems to be introduced in #11789.
